File: linkedLists/middleNode/middleNode.py
# O(n) time | O(1) space


class LinkedList:
    def __init__(self, value):
        self.value = value
        self.next = None


def middleNode(linkedList):
    slowNode = linkedList
    fastNode = linkedList
    while fastNode is not None and fastNode.next is not None:
        slowNode = slowNode.next
        fastNode = fastNode.next.next

    return slowNode


# Example usage:
if __name__ == "__main__":
    # Creating a linked list: 1 -> 2 -> 3 -> 4 -> 5
    head = LinkedList(1)
    head.next = LinkedList(2)
    head.next.next = LinkedList(3)
    head.next.next.next = LinkedList(4)
    head.next.next.next.next = LinkedList(5)

    middle = middleNode(head)
    print("Middle node value:", middle.value)  # Output: Middle node value: 3

File: linked_lists/middleNode.js
class LinkedList {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

exports.LinkedList = LinkedList;

function middleNode(linkedList) {
  let slowNode = linkedList
  let fastNode = linkedList

  while( fastNode !== null && fastNode.next !== null) {
    slowNode = slowNode.next
    fastNode = fastNode.next.next
  }
  return slowNode
}

// Exmaple usage:

const list = new LinkedList(1);
list.next = new LinkedList(2);
list.next.next = new LinkedList(3);
list.next.next.next = new LinkedList(4);
list.next.next.next.next = new LinkedList(5);
const middle = middleNode(list);

console.log(middle.value); // Output: 3